elastic/kibana
# [Uptime] [Synthetics Integration] Add script recorder tab for browser based monitors

## Summary

Fixes elastic/uptime#371

Adds the option to import scripts generated by the Elastic Synthetics Recorder.

<img width="402" alt="Screen Shot 2021-10-04 at 7 28 28 PM" src="https://user-images.githubusercontent.com/11356435/135938054-74b65aa1-ec80-434c-bd66-b482fd3ee898.png">

<img width="390" alt="Screen Shot 2021-10-04 at 7 29 20 PM" src="https://user-images.githubusercontent.com/11356435/135938288-6e4fe0ab-2ba9-47d4-b742-73fcfea2512f.png">
<img width="827" alt="Screen Shot 2021-10-04 at 7 29 28 PM" src="https://user-images.githubusercontent.com/11356435/135938289-40383e7a-2995-4f38-a988-6a16eaac3188.png">

_**Testing**_

**Setting up your e2e testing environment**
Overview: You will need to run a local version of kibana and package registry to test pending changes, but will use `elastic-package` to run a snapshot view of elasticsearch, fleet-server, and elastic-agent.  Since fleet server relies on kibana, we'll also use `elastic-package` to start a snapshot view of kibana exclusively for bootstrapping fleet-server, but will point our local version of kibana to a separate port. So we'll be running two versions of Kibana.
1. Stop all instances of elasticsearch or kibana
2. Checkout this PR. In `kibana.dev.yml` settings, add remove `kibana.index` and `xpack.task_manager.index` to remove legacy multitenancy features. This ensures that both our instances of Kibana can share saved objects related to fleet and agent. Also, add the following keys:
```
xpack.encryptedSavedObjects.encryptionKey: "12345678901234567890123456789012"
xpack.fleet.registryUrl: "http://localhost:8080"
server.port: 5602
```
4. Start your local kibana instance using `env BUILD_TS_REFS_DISABLE=true yarn kbn bootstrap && yarn start`
5. Update elastic package. It was recently updated to add support for the elastic-agent-complete image which contains support for running synthetics agent via heartbeat. Run `go get github.com/elastic/elastic-package`
6. Checkout https://github.com/elastic/integrations/pull/1778. Navigate to `packages/synthetics` in the repo directory. Run `elastic-package clean`, then `elastic-package build`.
7. Run `elastic-package stack up -d -v --services "elasticsearch,fleet-server,elastic-agent,package-registry" --version 8.0.0-SNAPSHOT` in the `packages/synthetics` directory. This will start up elasticsearch, fleet-server, elastic-agent, package registry AND kibana. This is because fleet-server requires kibana for bootstrapping.
8. Change the password to `kibana_system` user `curl -u elastic:changeme -X POST "http://localhost:9200/_security/user/kibana_system/_password?pretty" -H 'Content-Type: application/json' -d' { "password" : "changeme" } '`
9. Navigate to your local instance of kibana at **http://localhost:5602**

**Testing happy path**
1. Navigate to the synthetics integration at `fleet/integrations/synthetics-0.3.1/add-integration`
2. Add a policy name, then select the browser type from the dropdown
3. Select the "Script recorder" tab. 
4. Upload an inline script file. This should be a synthetics script starting with `step`. There are a few you can use in the synthetics repo in the `examples/inline` directory if you have it pulled down, or you can use the script recorder to actually create a script.
5. Save the integration and navigate to Uptime
6. Verify that the journey was ran

**Testing invalid file**
1. Navigate to the synthetics integration at `fleet/integrations/synthetics-0.3.1/add-integration`
2. Add a policy name, then select the browser type from the dropdown
3. Select the "Script recorder" tab. 
4. Upload anything besides a javascript file. Observe the error.

**Testing non inline journey**
1. Navigate to the synthetics integration at `fleet/integrations/synthetics-0.3.1/add-integration`
2. Add a policy name, then select the browser type from the dropdown
3. Select the "Script recorder" tab. 
4. Upload a .js file that does not start with step and is therefore not an inline journey. Observe the error.
